ARROW-12499: [C++][Compute] Add ScalarAggregateOptions to Any and All kernels

rok opened the pull request:
This is to resolve [ARROW-12499](https://issues.apache.org/jira/browse/ARROW-12499).

lidavidm reviewed (commented):
C++ parts look good to me.

Now that the kernel has options, should we add a Python convenience function in `pyarrow/compute.py`?

rok commented:
> C++ parts look good to me.

I'm thinking to if boolean checks would be better instead of counts?

> Now that the kernel has options, should we add a Python convenience function in `pyarrow/compute.py`?

Yeah definitely, next commit :)

lidavidm commented:
> > C++ parts look good to me.
> 
> I'm thinking to if boolean checks would be better instead of counts?

I debated whether to comment on this. I would slightly prefer the boolean check over the count but the intent is clear either way.

rok commented:
@lidavidm added the python test and switched from `null` count to `has_nulls` boolean. Also added a Python test so this should be good to review now.

lidavidm commented on python/pyarrow/tests/test_compute.py:
```diff
@@ -494,30 +494,40 @@ def test_min_max():
 def test_any():
     # ARROW-1846
+
+    options = pc.ScalarAggregateOptions(skip_nulls=False)
```
Hmm, I had thought that generally, for kernels with options, we add a wrapper in pyarrow/compute.py. But now that I look, this isn't true for many of the kernels. So having Python tests is enough.

rok commented:
Hey @thisisnic, @jonkeane, could you take a look at the [R part](https://github.com/apache/arrow/pull/10476/files#diff-9d33f036186210a1076f3adc590e0cc60a0b65556cefdbd15e6d26aa105f7d76) of this PR? It would be great to have an extra check :).

lidavidm reviewed (approved):
LGTM. I'm not quite sure what to make of the R tests so I'll let someone else comment on those.

jonkeane commented on r/tests/testthat/test-compute-aggregate.R:
```diff
@@ -383,33 +383,30 @@ test_that("value_counts", {
   data <- c(1:10, NA, NA)
 
-  expect_vector_equal(any(input > 5), data)
+  expect_vector_equal(any(input > 5, na.rm = TRUE), data)
```
Does this test fail without the `, na.rm = TRUE`?

rok replied:
It does indeed:
```R
Failure (test-compute-aggregate.R:388:3): any.Array and any.ChunkedArray
as.vector(x) not equal to `y`.
'is.NA' value mismatch: 0 in current 1 in target
Backtrace:
 1. expect_vector_equal(any(input > 5), data) test-compute-aggregate.R:388:2
 2. expect_as_vector(via_array, expected, ignore_attr, ...) helper-expectation.R:174:4
 3. testthat:::expect_fun(as.vector(x), y, ...) helper-expectation.R:24:2

Failure (test-compute-aggregate.R:388:3): any.Array and any.ChunkedArray
as.vector(x) not equal to `y`.
'is.NA' value mismatch: 0 in current 1 in target
Backtrace:
 1. expect_vector_equal(any(input > 5), data) test-compute-aggregate.R:388:2
 2. expect_as_vector(via_chunked, expected, ignore_attr, ...) helper-expectation.R:187:4
 3. testthat:::expect_fun(as.vector(x), y, ...) helper-expectation.R:24:2
```

rok replied:
I'm not sure how to read what happens in `expect_vector_equal`.

jonkeane replied:
[`expect_vector_equal` takes an expression and a vector](https://github.com/apache/arrow/blob/ed4f79c77e5d34d5996fa287df0934dd2f93a12a/r/tests/testthat/helper-expectation.R#L159-L195) and evaluates the expression in base R, then converts the vector converts it into an array and evaluates the expression again (which typically will call arrow compute functions), and finally converts the vector to a chunked array and does the same.

jonkeane replied:
So the errors you're seeing indicate that we've got a mismatch in behavior. I see that Ian has commented below about more robust tests + how to match the behavior (but I'm happy to add to it if that's unclear!)

rok replied:
Indeed @jonkeane it was a behavior issue. Should be fixed now.

jonkeane commented on r/tests/testthat/test-compute-aggregate.R:
```diff
@@ -383,33 +383,30 @@ test_that("value_counts", {
   data_logical <- c(TRUE, FALSE, TRUE, NA, FALSE)
   
-  expect_vector_equal(any(input), data_logical)
+  expect_vector_equal(any(input, na.rm = TRUE), data_logical)
```
Same question as above, this line and the line just below look like they are not testing the exact same thing, yeah? Am I missing something?

ianmcook replied:
This line and the one below are now identical. Let's please test all three cases (`na.rm` not specified, set to `FALSE`, and set to `TRUE`:
```suggestion
  expect_vector_equal(any(input), data_logical)
  expect_vector_equal(any(input, na.rm = FALSE), data_logical)
```

jonkeane reviewed (changes requested):
A few questions

ianmcook commented on r/tests/testthat/test-compute-aggregate.R:
```diff
@@ -383,33 +383,30 @@ test_that("value_counts", {
   expect_vector_equal(any(input < 1), data)
   expect_vector_equal(any(input < 1, na.rm = TRUE), data)
   
   data_logical <- c(TRUE, FALSE, TRUE, NA, FALSE)
```
To properly test that `any()` is handling `na.rm` consistent with the way base R does, we should test with a vector that does not have any `TRUE` in it. Maybe let's add another few lines of test code here to verify that, and perserve the existing tests too:
```suggestion
  data_logical <- c(TRUE, FALSE, TRUE, NA, FALSE)
  expect_vector_equal(any(input), data_logical)
  expect_vector_equal(any(input, na.rm = FALSE), data_logical)
  expect_vector_equal(any(input, na.rm = TRUE), data_logical)
  
  data_logical <- c(TRUE, FALSE, TRUE, NA, FALSE)
```

rok commented:
@lidavidm - I've introduced some small changes to c++ to enable desired behaviour so perhaps re-review is in order. Changes are all in [this commit.](https://github.com/apache/arrow/pull/10476/commits/69763766a7b44c66c8f9cce09845b3e9d1b6e218)

jorisvandenbossche reviewed (commented):
For the `and`/`or` kernels (the non-reducing variants of this), we have a separate "kleene" version of those kernels. So I am wondering: do we need that here as well?

Apparently a few months ago I though that this was needed, given ARROW-10291 / https://github.com/apache/arrow/pull/8294#issuecomment-701357437, but to be honest I don't fully understand my own comment there any more .. ;)

It seems that the current PR implements the "kleene" version, and I am not fully sure how a non-kleene version would behave otherwise and if this would actually ever be useful. 

In any case we should probably describe the behaviour somewhere more explicitly.

---

Do we want any/all to follow the `min_count` parameter of the ScalarAggregateOptions? If not, we should probably document that it is being ignored.

rok commented on cpp/src/arrow/compute/kernels/aggregate_basic.cc:
```diff
@@ -166,32 +168,48 @@ struct BooleanAnyImpl : public ScalarAggregator {
-  Status Finalize(KernelContext*, Datum* out) override {
-    out->value = std::make_shared<BooleanScalar>(this->any);
+  Status Finalize(KernelContext* ctx, Datum* out) override {
+    if (!options.skip_nulls && !this->any && this->has_nulls) {
```
Indeed this is to to make the PR "kleen" to match [R behavior](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/any). Meanwhile [Pandas' any is non-kleen](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.any.html).

```
>>> import pandas as pd
>>> pd.Series([None, None]).any(skipna=True)
False
>>> pd.Series([None, None]).any(skipna=False)
>>> 
```

We have three options IMO:
- Revert to non-kleene c++ behaviour and add a small fix to R
- Add kleen any/all kernels and route in R depending on flags
- Keep c++ as is and add a fix to Python (that could introduce a lot of new unvanted logic)

jorisvandenbossche replied:
> Meanwhile Pandas' any is non-kleen.

The pandas `any`/`all` methods were broken for object dtype (and since numpy doesn't support nulls in its boolean dtype, whenever you have missing values, you have object dtype), so best not to use that as a reference (see eg https://github.com/pandas-dev/pandas/issues/27709)

rok replied:
So how about then just making this kernel "kleene" and just document that fact?

rok replied:
Shall I:
1. Document the behavior as kleen?
2. Rather revert the change and first implement `al_kleen/any_kleen` ([ARROW-10291](https://issues.apache.org/jira/browse/ARROW-10291)) and then map R to those for `skip_nulls==False`?

jorisvandenbossche replied:
I am not sure there is a good use case for a non-kleene version, so I am fine with just documenting for now that the behaviour follows Kleene logic (so is the reducing version of and/or_kleene)

rok replied:
I documented the new behavior in [compute.rst](https://github.com/apache/arrow/pull/10476/files#diff-ce5b94577014735990903d3d03bd4ea4b8c8e6d32f5227592e60b7dd6a912d59) and [api_aggregate.h](https://github.com/apache/arrow/pull/10476/files#diff-694a635f185cf0d22dd6eefd91acedb48ed0bf71b15fd591aa6e7c508ba4cb09).

jorisvandenbossche commented on cpp/src/arrow/compute/api_aggregate.h:
```diff
@@ -201,30 +201,42 @@ Result<Datum> MinMax(
-/// to true and false otherwise. Null values are skipped.
+/// to true and false otherwise. Null values are ignored by default.
+/// If null values are taken into account by setting ScalarAggregateOptions
+/// parameter skip_nulls = false then kleen logic is applied.
```
```suggestion
/// parameter skip_nulls = false then Kleene logic is used.
```

Maybe also add something like "see OrKleene for details on Kleene logic"

rok replied:
I think it KleeneOr. Added.

jorisvandenbossche commented on cpp/src/arrow/compute/api_aggregate.h:
```diff
@@ -201,30 +201,42 @@ Result<Datum> MinMax(
-/// to true and false otherwise. Null values are skipped.
+/// to true and false otherwise. Null values are ignored by default.
+/// If null values are taken into account by setting ScalarAggregateOptions
+/// parameter skip_nulls = false then kleen logic is applied.
```
```suggestion
/// parameter skip_nulls = false then Kleene logic is used.
```

jorisvandenbossche commented on docs/source/cpp/compute.rst:
```diff
@@ -213,18 +213,23 @@ Aggregations
+* \(5) Output is Int64, UInt64 or Float64, depending on the input type.
 
-* \(4) Output is Int64, UInt64 or Float64, depending on the input type.
+.. _Kleene logic: https://en.wikipedia.org/wiki/Three-valued_logic#Kleene_and_Priest_logics
```
This target is already defined below as well, I think?

rok replied:
Removed. I was not sure how scopes in rst work :).

jorisvandenbossche reviewed (commented):
Can you also update the docstrings which are defined in aggregate_basic.cc ?

jorisvandenbossche commented on cpp/src/arrow/compute/api_aggregate.h:
```diff
@@ -201,30 +201,44 @@ Result<Datum> MinMax(
+/// to true and false otherwise. Null values are ignored by default.
+/// If null values are taken into account by setting ScalarAggregateOptions
+/// parameter skip_nulls = false then Kleene logic is used.
+/// See KleeneOr for more details on Kleene logic.
```
```suggestion
/// See KleeneAnd for more details on Kleene logic.
```

jorisvandenbossche commented:
Can you take a look at my comment from above, which I think is not yet addressed:

> Can you also update the docstrings which are defined in aggregate_basic.cc ?

rok commented:
> Can you take a look at my comment from above, which I think is not yet addressed:
> 
> > Can you also update the docstrings which are defined in aggregate_basic.cc ?

Oh sorry, I missed that one. Added in [last commit](https://github.com/apache/arrow/pull/10476/commits/c5976a94287cd5a0fd122fc4cbe56be93719b293).

rok commented:
Ping :)

rok commented:
@jorisvandenbossche @jonkeane @ianmcook Does this need another round of reviews?

rok commented:
Thanks all!
